api/server/sdk/alerts.go

```diff
@@ -0,0 +1,157 @@
+			options = append(options,
+				alerts.NewFlagCheckOption(opt.GetIsCleared()))
+		case *api.SdkAlertsOption_TimeSpan:
+			options = append(options,
+				alerts.NewTimeSpanOption(
+					prototime.TimestampToTime(opt.GetTimeSpan().GetStartTime()),
+					prototime.TimestampToTime(opt.GetTimeSpan().GetEndTime())))
+		case *api.SdkAlertsOption_CountSpan:
+			options = append(options,
+				alerts.NewCountSpanOption(
+					opt.GetCountSpan().GetMinCount(),
+					opt.GetCountSpan().GetMaxCount()))
+		}
+	}
+
+	return options
+}
+
+func getFilters(queries []*api.SdkAlertsQuery) []alerts.Filter {
+	var filters []alerts.Filter
+
+	// range over all queries
+	for _, x := range queries {
+		switch x.GetQuery().(type) {
+		case *api.SdkAlertsQuery_ResourceTypeQuery:
+			q := x.GetResourceTypeQuery()
+			filters = append(filters,
+				alerts.NewResourceTypeFilter(
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		case *api.SdkAlertsQuery_AlertTypeQuery:
+			q := x.GetAlertTypeQuery()
+			filters = append(filters,
+				alerts.NewAlertTypeFilter(
+					q.AlertType,
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		case *api.SdkAlertsQuery_ResourceIdQuery:
+			q := x.GetResourceIdQuery()
+			filters = append(filters,
+				alerts.NewResourceIDFilter(
+					q.ResourceId,
+					q.AlertType,
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		}
+	}
+
+	return filters
+}
+
+// Enumerate implements api.OpenStorageAlertsServer for alertsServer.
+// Input context should ideally have a deadline, in which case, a
+// graceful exit is ensured within that deadline.
+func (g *alertsServer) Enumerate(ctx context.Context,
+	request *api.SdkAlertsEnumerateRequest) (*api.SdkAlertsEnumerateResponse, error) {
+
+	// if input has deadline, ensure graceful exit within that deadline.
+	deadline, ok := ctx.Deadline()
+	var cancel context.CancelFunc
+	if ok {
+		// create a new context that will get done on deadline
+		ctx, cancel = context.WithTimeout(ctx, deadline.Sub(time.Now()))
+		defer cancel()
+	}
+
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+
+	resp := new(api.SdkAlertsEnumerateResponse)
+	var mu sync.Mutex
+
+	queries := request.GetQueries()
+	if queries == nil {
+		return nil, status.Error(codes.InvalidArgument, "input is nil")
+	}
+
+	filters := getFilters(queries)
+
+	// spwan err-group process for each independent query.
+	// collect output using mutex.
+	group.Go(func() error {
+		if out, err := g.reader.Enumerate(filters...); err != nil {
+			return err
+		} else {
+			mu.Lock()
+			resp.Alert = append(resp.Alert, out...)
+			mu.Unlock()
+			return nil
+		}
+	})
+
+	// wait for err-group processes to be done
+	go func() {
+		errChan <- group.Wait()
+	}()
+
+	// wait only as long as context deadline allows
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return nil, err
+		} else {
+			return resp, nil
+		}
+	case <-ctx.Done():
+		return nil, status.Error(codes.DeadlineExceeded,
```

Comment:
Can the context be Done() with success? or is ctx.Done() only called for errors?

Reply:
if the context has deadline, or is cancellable and cancelled, only then context is done. otherwise, this case with block forever. context.Background(), for instance, is never done.